src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -49924,6 +49924,240 @@ static object F((I<object>, I<object?>)? x, (I<object?>, I<object>) y)
                 Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "y").WithArguments("(I<object?>, I<object>)", "(I<object>, I<object?>)").WithLocation(10, 21));
         }
 
+        [Fact, WorkItem(51975, "https://github.com/dotnet/roslyn/issues/51975")]
+        public void NullCoalescingOperator_09()
+        {
+            var source =
+@"C? c = new C();
+D d = c ?? new D();
+
+d.ToString();
+
+class C {}
+
+class D
+{
+    public static implicit operator D?(C c) => default;
+}
+";
+
+            var comp = CreateCompilation(source, options: WithNullableEnable(TestOptions.ReleaseExe));
+            comp.VerifyDiagnostics(
+                // (2,7): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                // D d = c ?? new D();
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "c ?? new D()").WithLocation(2, 7),
+                // (4,1): warning CS8602: Dereference of a possibly null reference.
+                // d.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "d").WithLocation(4, 1)
+            );
+        }
+
+        [Fact, WorkItem(51975, "https://github.com/dotnet/roslyn/issues/51975")]
+        public void NullCoalescingOperator_10()
+        {
+            var source =
+@"C? c = new C();
+D d = c ?? new D();
+
+d.ToString();
+
+class C
+{
+    public static implicit operator D?(C c) => default;
+}
+
+class D {}
+";
+
+            var comp = CreateCompilation(source, options: WithNullableEnable(TestOptions.ReleaseExe));
+            comp.VerifyDiagnostics(
+                // (2,7): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                // D d = c ?? new D();
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "c ?? new D()").WithLocation(2, 7),
+                // (4,1): warning CS8602: Dereference of a possibly null reference.
+                // d.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "d").WithLocation(4, 1)
+            );
+        }
+
+        [Fact, WorkItem(51975, "https://github.com/dotnet/roslyn/issues/51975")]
+        public void NullCoalescingOperator_11()
```

Comment:
For next time: Adding some hint to the test (name or body) about what is special would save reviewers time in playing game of difference...